### PR TITLE
Move to evalutate scala reflect only once

### PR DESCRIPTION
**Pull request by steveroy0226**
This PR makes it so that the scala-reflection is evaluated only once. This leads to a huge performance improvement.

Before this change, a sample Dataflow workload had this data:
- Current vCPUs: 12
- Total vCPU time: 1.101 vCPU hr
- Current memory: 10.8 GB
- Total memory time: 0.991 GB hr
- Current HDD PD: 75 GB
- Total HDD PD time: 6.881 GB hr
- Elapsed time: 8 min 44 sec

After this change, the improvement to the same Dataflow workload is as follows:
- Current vCPUs: 4
- Total vCPU time: 0.168 vCPU hr
- Current memory: 3.6 GB
- Total memory time: 0.151 GB hr
- Current HDD PD: 25 GB
- Total HDD PD time: 1.048 GB hr
- Elapsed time: 4 min 7 sec

Also, this PR adds a change to properly return a meaningful message if the given filed doesn't exist in the schema.

### CODE REVIEW

**Review by benkonz — APPROVED** [review]
I don't really understand the inner-workings of some of what's going on here, but LGTM! Maybe wait for a second opinion before merging

File: elitzur-avro/src/main/scala/com/spotify/elitzur/converters/avro/dynamic/DynamicAccessorCompanion.scala
Comment (catherinejelder) [inline elitzur-avro/src/main/scala/com/spotify/elitzur/converters/avro/dynamic/DynamicAccessorCompanion.scala:43]: I'm not sure this link is particularly helpful here - it's oriented around fixing `NotSerializableException`s, which doesn't seem particularly relevant

**Review by catherinejelder — COMMENTED** [review]
I also don't yet understand why this gets us a performance gain, but I'd like to understand - maybe you could explain it at monday's standup?
it doesn't look harmful either, but I'd like to understand better before approving :)

File: elitzur-avro/src/main/scala/com/spotify/elitzur/converters/avro/dynamic/DynamicAccessorCompanion.scala
Comment (steveroy0226) [reply elitzur-avro/src/main/scala/com/spotify/elitzur/converters/avro/dynamic/DynamicAccessorCompanion.scala:43]: I agree, will remove. I was generally avoiding using `@transient lazy` in the codebase and, at the time, felt adding an explanation why this was required was a good practice. But it seems out-of-place now.

**Comment by steveroy0226** [comment]
> I also don't yet understand why this gets us a performance gain, but I'd like to understand - maybe you could explain it at monday's standup?
> it doesn't look harmful either, but I'd like to understand better before approving :)

Yup! I will share in our internal channel the profiling results from before and after the change as well. Those profiling charts will probably be a lot better at conveying the actual changes at the JVM level.

**Review by catherinejelder — APPROVED** [review]
thanks for the explanation, this is a bit clearer to me now!

**Pull request merged**
